File: app/src/main/java/org/adaway/tile/AdBlockingTileService.java
package org.adaway.tile;

import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;

import androidx.lifecycle.LiveData;

import org.adaway.AdAwayApplication;
import org.adaway.model.adblocking.AdBlockModel;
import org.adaway.model.error.HostErrorException;
import org.adaway.util.AppExecutors;

import java.util.concurrent.atomic.AtomicBoolean;

import static android.service.quicksettings.Tile.STATE_ACTIVE;
import static android.service.quicksettings.Tile.STATE_INACTIVE;

import timber.log.Timber;

/**
 * This class is a {@link TileService} to toggle ad-blocking.
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
public class AdBlockingTileService extends TileService {
    private final AtomicBoolean toggling = new AtomicBoolean(false);

    @Override
    public void onTileAdded() {
        boolean adBlocked = getModel().isApplied().getValue() == Boolean.TRUE;
        updateTile(adBlocked);
    }

    @Override
    public void onStartListening() {
        LiveData<Boolean> applied = getModel().isApplied();
        applied.observeForever(this::updateTile);
    }

    @Override
    public void onStopListening() {
        LiveData<Boolean> applied = getModel().isApplied();
        applied.removeObserver(this::updateTile);
    }

    @Override
    public void onClick() {
        AppExecutors.getInstance()
                .diskIO()
                .execute(this::toggleAdBlocking);
    }

    private void updateTile(boolean adBlocked) {
        Tile tile = getQsTile();
        // The tile is null when the system has unbound the service (it may happen between
        // a click that triggered an async toggle and the toggle completing).
        if (tile == null) {
            return;
        }
        tile.setState(adBlocked ? STATE_ACTIVE : STATE_INACTIVE);
        tile.updateTile();
    }

    private void toggleAdBlocking() {
        if (this.toggling.get()) {
            return;
        }
        AdBlockModel model = getModel();
        boolean wasApplied = model.isApplied().getValue() == Boolean.TRUE;
        try {
            this.toggling.set(true);
            Timber.i("Tile: user toggle (currently %s).", wasApplied ? "ACTIVE" : "INACTIVE");
            if (wasApplied) {
                model.revert();
            } else {
                model.apply();
            }
        } catch (HostErrorException e) {
            Timber.w(e, "Failed to toggle ad-blocking.");
        } finally {
            this.toggling.set(false);
            // Whatever happened (success, failure, exception), force the tile back in
            // sync with the LiveData so a failed start doesn't leave the tile stuck on
            // an optimistic "ACTIVE" state.
            updateTile(model.isApplied().getValue() == Boolean.TRUE);
        }
    }

    private AdBlockModel getModel() {
        return ((AdAwayApplication) getApplication()).getAdBlockModel();
    }
}

File: app/src/main/java/org/adaway/tile/AdBlockingTileService.kt
package org.adaway.tile

import android.service.quicksettings.Tile.STATE_ACTIVE
import android.service.quicksettings.Tile.STATE_INACTIVE
import android.service.quicksettings.TileService
import org.adaway.AdAwayApplication
import org.adaway.model.adblocking.AdBlockModel
import org.adaway.model.error.HostErrorException
import org.adaway.util.CoroutineDispatchers
import timber.log.Timber
import java.util.concurrent.atomic.AtomicBoolean

class AdBlockingTileService : TileService() {
    private val toggling = AtomicBoolean(false)

    override fun onTileAdded() {
        updateTile(model.isApplied.value == true)
    }

    override fun onStartListening() {
        model.isApplied.observeForever(::updateTile)
    }

    override fun onStopListening() {
        model.isApplied.removeObserver(::updateTile)
    }

    override fun onClick() {
        CoroutineDispatchers.ioExecutor().execute(::toggleAdBlocking)
    }

    private fun updateTile(adBlocked: Boolean) {
        qsTile?.let { tile ->
            tile.state = if (adBlocked) STATE_ACTIVE else STATE_INACTIVE
            tile.updateTile()
        }
    }

    private fun toggleAdBlocking() {
        if (toggling.get()) {
            return
        }
        val model = model
        try {
            toggling.set(true)
            if (model.isApplied.value == true) {
                model.revert()
            } else {
                model.apply()
            }
        } catch (exception: HostErrorException) {
            Timber.w(exception, "Failed to toggle ad-blocking.")
        } finally {
            toggling.set(false)
        }
    }

    private val model: AdBlockModel
        get() = (application as AdAwayApplication).adBlockModel
}
